File: src/test/java/de/flapdoodle/generics/selftypes/j/SampleTest.java
package de.flapdoodle.generics.selftypes.j;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.*;

class SampleTest {

  @Test
  void callChangeName() {
    Sample testee = new Sample("foo", new ArrayList<>());

    Sample result = testee.changeName("bar");

    assertEquals("bar", result.name());
  }

  @Test
  void typeOfChildren() {
    Sample testee = new Sample("foo", Arrays.asList(new Sample("blub", new ArrayList<>())));

    List<? extends SampleType> children = testee.children();

    List<? extends SampleType> result = children.stream()
        .map(s -> changeName(s, "bar"))
        .collect(Collectors.toList());

    assertEquals(1, result.size());
    assertEquals("bar", result.get(0).name());
  }

  <T extends SampleType<T>> T changeName(T instance, String name) {
    return instance.changeName(name);
  }
}

File: src/test/kotlin/de/flapdoodle/generics/selftypes/k/SampleTest.kt
package de.flapdoodle.generics.selftypes.k

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.util.*
import java.util.stream.Collectors
import kotlin.streams.toList

class SampleTest {
  @Test
  fun callChangeName() {
    val testee = Sample("foo", ArrayList())

    val result = testee.changeName("bar")

    assertEquals("bar", result.name())
  }

  @Test
  fun typeOfChildren() {
    val testee = Sample("foo", Arrays.asList(Sample("blub", ArrayList())))

    val children = testee.children()

    val result = children.stream()
        .map { s -> changeName(s, "bar") }
        .toList()

    assertEquals(1, result.size)
    assertEquals("bar", result[0].name())
  }

  fun <T : SampleType<*>> changeName(instance: T, name: String): T {
    @Suppress("UNCHECKED_CAST")
    return instance.changeName(name) as T
  }

}